### ISSUE TITLE
linter: `explicit-function-return-type` `allowTypedFunctionExpressions` does not exempt function properties in contextually-typed object literals

### ISSUE THREAD

**Issue by daikieng** [oxc-project/oxc#24792]
### What version of Oxlint are you using?

1.75.0 (also reproduces on 1.71.0)

### What command did you run?

`oxlint -c .oxlintrc.json`

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
{
  "plugins": ["typescript"],
  "rules": {
    "typescript/explicit-function-return-type": [
      "error",
      {
        "allowExpressions": false,
        "allowTypedFunctionExpressions": true,
        "allowHigherOrderFunctions": true
      }
    ]
  }
}
```

### What happened?

**What happened?**

Repro (`src/typed_expr.ts`):

```ts
interface Repo {
  count: () => Promise<number>;
}

export const buildRepo = (): Repo => ({
  count: async () => 0, // line 6
});

export function untyped(x) { // line 9
  return x;
}
```

oxlint reports **2 errors**, including line 6:

```
src/typed_expr.ts:6:3: error typescript(explicit-function-return-type): Missing return type on function.
src/typed_expr.ts:9:8: error typescript(explicit-function-return-type): Missing return type on function.
```

typescript-eslint (8.18.2) with the identical options reports **only 1 error** (line 9). Line 6 is exempted by `allowTypedFunctionExpressions`: the arrow function's type is fully determined by the contextual type (`Repo`, via the enclosing arrow's return type annotation).

Returning an object literal that implements an annotated interface is idiomatic TypeScript (factory functions, DI providers, config objects, test doubles, etc.), where the contextual type fully determines and checks each property's function signature. Since typescript-eslint exempts these by design, the divergence produces widespread false positives in existing codebases and makes the rule hard to adopt as a drop-in ESLint replacement.

Possibly related: #12704 was resolved by aligning the option defaults (#12718), but contextually-typed object literal properties still diverge even with the option explicitly enabled.

**Expected behavior**:
Line 6 should not be reported when `allowTypedFunctionExpressions` is enabled, matching typescript-eslint.

Happy to attempt a fix PR if the expected behavior is confirmed.

**Comment by daikieng** [comment]
@camc314 
 I noticed you've assigned yourself to this.
If you haven't started on it yet, I'd be happy to take a shot at a fix PR, as I'd like to start contributing to oxc. If you're already on it, no worries at all!

**Comment by camc314** [comment]
i've got a draft PR up - i think there's some edge cases missing, feel free to help get it across the line !

**Comment by daikieng** [comment]
@camc314 
Sure, thanks!
